python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+        }
+
+    def create_dependency_graph(self):
+        """Create a dependency graph using the nodes and corresponding inputs"""
+        for node_name, curr_node in self.nodes.items():
+            if not curr_node.is_input_node:
+                for input_node in curr_node.inputs:
+                    self.nodes[input_node].dependent_children.append(curr_node)
+
+    def get_output_nodes(self) -> List[LayerNode]:
+        """Get a list of all output nodes"""
+        # Nodes which do not have outgoing edges (no downstream dependencies), except the feature nodes
+        return list(filter(lambda node: (not node.is_input_node) and (len(node.dependent_children) == 0),
+                           self.nodes.values()))
+
+    def topological_sort(self) -> List[LayerNode]:
+        """
+        Get the execution order of nodes such that all inputs to a layer is available for forward prop
+        This method also ensures there are no cycles in the DAG. If a cycle is found, throws `CycleFoundException`
+        """
+        order = []
+        path = set()
+        visited = set()
+
+        def dfs(node: LayerNode):
+            nonlocal order, visited, path
+            visited.add(node.name)
+            for child in node.dependent_children:
+                if child.name in path:
+                    raise CycleFoundException("Cycle found in graph. Only DAGs are supported")
+                if child.name not in visited:
+                    path.add(child.name)
+                    dfs(child)
+                    path.remove(child.name)
+            order.append(node)
+
+        for node in self.nodes.values():
+            if node.name not in visited:
+                path.add(node.name)
+                dfs(node)
+                path.remove(node.name)
+
+        return order[::-1]
+
+
+class AutoDagNetwork(keras.Model):
+    """DAG model architecture dynamically inferred from model config that maps features -> logits"""
+    INPUTS = "inputs"
+    INPUTS_AS_LIST = "aslist"
+    LAYERS_IDENTIFIER = "layers"
+    LAYER_TYPE = "type"
+    NAME = "name"
+    OP_IDENTIFIER = "op"
+    LAYER_KWARGS = "args"
+
+    def __init__(
+            self,
+            model_config: dict,
+            feature_config: FeatureConfig,
+            file_io: FileIO,
+            **kwargs
+    ):
+        """
+        Initialize a dense neural network layer
+
+        Parameters
+        ----------
+        model_config: dict
+            Dictionary defining the dense architecture spec
+        feature_config: FeatureConfig
+            FeatureConfig defining how each input feature is used in the model
+        file_io: FileIO
+            File input output handler
+        """
+        super().__init__(**kwargs)
+
+        self.file_io: FileIO = file_io
+        self.model_config = model_config
+        self.feature_config = feature_config
+
+        # Get all available layers (including keras layers)
+        self.available_layers = get_layer_subclasses()
```

Comment:
@arvindsrikantan Worth making this an argument to the constructor with the default being `get_layer_subclasses()`